tests/fixture/testnet/local/network.go

```diff
@@ -0,0 +1,644 @@
+		return nil, err
+	}
+
+	if len(rootDir) == 0 {
+		// Use the default root dir
+		var err error
+		rootDir, err = GetDefaultRootDir()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Ensure creation of the root dir
+	if err := os.MkdirAll(rootDir, perms.ReadWriteExecute); err != nil {
+		return nil, fmt.Errorf("failed to create root network dir: %w", err)
+	}
+
+	// Determine the network path and ID
+	var (
+		networkDir string
+		networkID  uint32
+	)
+	if network.Genesis != nil && network.Genesis.NetworkID > 0 {
+		// Use the network ID defined in the provided genesis
+		networkID = network.Genesis.NetworkID
+	}
+	if networkID > 0 {
+		// Use a directory with a random suffix
+		var err error
+		networkDir, err = os.MkdirTemp(rootDir, fmt.Sprintf("%d.", network.Genesis.NetworkID))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create network dir: %w", err)
+		}
+	} else {
+		// Find the next available network ID based on the contents of the root dir
+		var err error
+		networkID, networkDir, err = FindNextNetworkID(rootDir)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Setting the network dir before populating config ensures the
+	// nodes know where to write their configuration.
+	network.Dir = networkDir
+
+	if err := network.PopulateLocalNetworkConfig(networkID, nodeCount, keyCount); err != nil {
+		return nil, err
+	}
+
+	if err := network.WriteAll(); err != nil {
+		return nil, err
+	}
+	if _, err := fmt.Fprintf(w, "Starting network %d @ %s\n", network.Genesis.NetworkID, network.Dir); err != nil {
+		return nil, err
+	}
+	if err := network.Start(w); err != nil {
+		return nil, err
+	}
+	if _, err := fmt.Fprintf(w, "Waiting for all nodes to report healthy...\n\n"); err != nil {
+		return nil, err
+	}
+	if err := network.WaitForHealthy(ctx, w); err != nil {
+		return nil, err
+	}
+	if _, err := fmt.Fprintf(w, "\nStarted network %d @ %s\n", network.Genesis.NetworkID, network.Dir); err != nil {
+		return nil, err
+	}
+	return network, nil
+}
+
+// Read a network from the provided directory.
+func ReadNetwork(dir string) (*LocalNetwork, error) {
+	network := &LocalNetwork{Dir: dir}
+	if err := network.ReadAll(); err != nil {
+		return nil, fmt.Errorf("failed to read local network: %w", err)
+	}
+	return network, nil
+}
+
+// Stop the nodes of the network configured in the provided directory.
+func StopNetwork(dir string) error {
+	network, err := ReadNetwork(dir)
+	if err != nil {
+		return err
+	}
+	return network.Stop()
+}
+
+// Ensure the network has the configuration it needs to start.
+func (ln *LocalNetwork) PopulateLocalNetworkConfig(networkID uint32, nodeCount int, keyCount int) error {
+	if len(ln.Nodes) > 0 && nodeCount > 0 {
+		return errInvalidNodeCount
+	}
+	if len(ln.FundedKeys) > 0 && keyCount > 0 {
+		return errInvalidKeyCount
+	}
+
+	if nodeCount > 0 {
+		// Add the specified number of nodes
+		nodes := []*LocalNode{}
```

Comment:
Not sure we really care to pre-alloc stuff like this in the testing harness... But we can also pre-alloc here.